### ISSUE TITLE
Implement Automatic Shader Uniform Display

### ISSUE THREAD

**Issue by cannontwo** [cannontwo/cannon#60]
`VertexShader` and `FragmentShader` should be able to scan their input strings and extract uniforms, then write them to ImGUI automatically. In order to do this, they will keep a map of string names to `ShaderUniform` shared pointers. `ShaderUniform` will be polymorphic, with derived classes implementing functionality specific to each type of uniform supported by `ShaderProgram.set_uniform`. When `set_uniform` is called, a method on the appropriate `ShaderUniform` object should be called instead of the direct setting currently implemented. Then, in `ShaderProgram.activate`, a virtual method on `ShaderUniform` to set the actual uniform on the OpenGL shader object should be called. 

In summary, `ShaderUniform` defines two methods:
1. `void write_imgui()`
2. `void set_uniform()`

Each subclass must also implement a method `update_value` which has a signature specific to its uniform type, which will be called in the `set_uniform` method of `ShaderProgram` by dynamic casting the `ShaderUniform` shared pointer.
